gift-cards/security.md

```diff
@@ -0,0 +1,240 @@
+
+```java
+public class Oauth2RestTemplateConfiguration {
+    //...
+    @Bean
+    @ConfigurationProperties("vas-payment-api.oauth2.client")
+    protected ClientCredentialsResourceDetails oAuthDetails() {
+        return new ClientCredentialsResourceDetails();
+    }
+
+    @Bean
+    protected RestTemplate restTemplate() {
+        var restTemplate = new OAuth2RestTemplate(oAuthDetails());
+        restTemplate.setInterceptors(ImmutableList.of(externalRequestInterceptor()));
+        restTemplate.setRequestFactory(httpRequestFactory());
+        return restTemplate;
+    }
+    //...
+}
+```
+
+## HMAC
+
+The API also requires HMAC authentication to be present in a request. In this
+client the HMAC value is automatically calculated by
+[HmacSignatureBuilder.java][hmac-signature-builder] and added to all outgoing
+requests in [ExternalRequestInterceptor.java][external-request-interceptor]
+
+HMAC is implemented using SHA-512 secure hash algorithm.
+
+Expected `Hmac` header format is:
+
+```text
+HmacSHA512 <user>:<nonce>:<digest>
+```
+
+where `digest` is a Base64 formatted HMAC SHA512 digest of the following string:
+
+```text
+METHOD\n
+RESOURCE\n
+USER\
+NONCE\n
+DATE\n
+PAYLOAD\n
+```
+
+{:.table .table-striped}
+|     Required     | Field                   | Description                                                                                                                     |
+| :--------------: | :---------------------- | :------------------------------------------------------------------------------------------------------------------------------ |
+| {% icon check %} | `METHOD`     | The requested method (in upper case)                                                                                                                                 |
+| {% icon check %} | `RESOURCE` | The path to desired resource (without hostname and any query parameters)
+| {% icon check %} | `NONSE` | A unique value for each request ([UUID][uuid]
+|                  | `DATE` | Same as `Transmission-Time` if provided as seperate header. Uses [ISO8601 standard][iso-8601]
+|                  | `PAYLOAD` | The body of request
+
+Example request:
+
+```bash
+curl -X POST \
+  https://stage-evc.payex.com/payment-api/api/payments/payment-account/balance \
+  -H 'Accept: */*' \
+  -H 'Agreement-Merchant-Id: XXX' \
+  -H 'Authorization: Bearer XXX' \
+  -H 'Hmac: HmacSHA512 user:21a0213e-30eb-85ab-b355-a310d31af30e:oY5Q5Rf1anCz7DRm3GyWR0dvJDnhl/psylfnNCn6FA0NOrQS3L0fvyUsQ1IQ9gQPeLUt9J3IM2zwoSfZpDgRJA==' \
+  -H 'Transmission-Time: 2019-06-18T09:19:15.208257Z' \
+  -H 'Session-Id: e0447bd2-ab64-b456-b17b-da274bb8428e' \
+  -d '{
+    "accountIdentifier": {
+        "accountKey": "7013369000000000000",
+        "cvc": "123",
+        "expiryDate": "2019-12-31",
+        "instrument": "GC"
+    }
+}'
+```
+
+In this example `USER` is user and `SECRET` is secret.
+
+The plain text string to `digest` would then look like the following:
+
+```text
```

Comment:
Why is this `text` when it could be `http`?

Reply:
It looks like HTTP, but isn't. I did the same mistake myself. Read the previous paragraph and you realize it is just the `string` content of what will be encrypted and become the HMAC.